lib/elixir/lib/enum.ex

```diff
@@ -3214,6 +3214,51 @@ defmodule Enum do
     |> :lists.reverse()
   end
 
+  @doc """
+  Zips two collections together by passing the corresponding element from each collection to the
+  provided fun. Returns a new list that contains the result of that fun for each pair of elements.
+
+  The zipping finishes as soon as any enumerable in the given collection completes.
+
+  ## Examples
+
+      iex> Enum.zip_with([1, 2], [3, 4], fn x, y -> x + y end)
+      [4, 6]
+
+      iex> Enum.zip_with([1, 2], [3, 4, 5, 6], fn x, y -> x + y end)
+      [4, 6]
+
+      iex> Enum.zip_with([1, 2, 5, 6], [3, 4], fn x, y -> x + y end)
+      [4, 6]
+  """
+  @spec zip_with(t, t, (term, term -> term)) :: [term]
+  def zip_with(enumerable1, enumerable2, fun)
+      when is_list(enumerable1) and is_list(enumerable2) do
+    zip_list(enumerable1, enumerable2, fun)
+  end
+
+  def zip_with(enumerable1, enumerable2, fun) do
+    zip_with([enumerable1, enumerable2], fun)
```

Comment:
The following example will fail:

    iex> Enum.zip_with([1, 2, 5, 6], 3..4, fn x, y -> x + y end)

That's because we call `zip_with` and `zip_with` with pass a list to the `fn`.  You can fix it by calling it like this:

```suggestion
    zip_with([enumerable1, enumerable2], &apply(fun, &1))
```

You can also take this approach on `Stream.zip_with/3` so it calls `fun` with two args.

Reply:
Btw, please add tests mixing lists with ranges and similar.

Reply:
Nice spot! Will do.

What's your thoughts on having zip_with/2's zipping fun receive a list instead?
Essentially turn this line:  https://github.com/elixir-lang/elixir/pull/10419/files#diff-6881431a92cd4e3ea0de82bf2338f8eaR3797
into this:
```elixir 
  defp zip_list([h1 | next1], [h2 | next2], fun) do
   # pass a list here to fun
    [fun.([h1, h2]) | zip_list(next1, next2, fun)]
  end
```

I feel like it will be more consistent